src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/TextCompletionItemProvider.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.VisualStudio.Editor.Razor;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+
+namespace Microsoft.CodeAnalysis.Razor.Completion
+{
+    internal class TextCompletionItemProvider : RazorCompletionItemProvider
+    {
+        private static readonly string TextTag = "text";
+        private static readonly HashSet<string> ElementCommitCharacters = new HashSet<string>{" ", ">"};
+        private readonly HtmlFactsService _htmlFactsService;
+
+        public TextCompletionItemProvider(HtmlFactsService htmlFactsService)
+        {
+            if (htmlFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(htmlFactsService));
+            }
+
+            _htmlFactsService = htmlFactsService;
+        }
+
+        public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorSyntaxTree syntaxTree, TagHelperDocumentContext tagHelperDocumentContext, SourceSpan location)
+        {
+            var completions = new List<RazorCompletionItem>();
+            if (syntaxTree == null)
+            {
+                return completions;
```

Comment:
When you're purposefully returning empty types with the intent that they contain nothing be explicit and use empty objects. The value in doing this is it helps readers understand that the intent in the statement (nothing to be returned); otherwise as a reader you need to scan up the method to see if anything was added.
```suggestion
                return Array.Empty<RazorCompletionItem>();
```